src/linear_algebra/orientation.lean

```diff
@@ -0,0 +1,376 @@
+/-
+Copyright (c) 2021 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Myers
+-/
+import linear_algebra.determinant
+
+/-!
+# Orientations of modules
+
+This file defines orientation of modules.
+
+## Main definitions
+
+* `orientation` is a type class for a choice of orientation of a module.  An orientation may
+be associated with an alternating map or with a basis.
+
+## Implementation notes
+
+`orientation` is defined for an arbitrary index type, but the main intended use case is when
+that index type is a `fintype` and there exists a basis of the same cardinality.
+
+Although `orientation` is a type class, many lemmas have an `orientation` as an explicit
+argument, in order to talk about how two orientations are related or to state results in
+contexts where a particular orientation is not being chosen as canonical.
+
+## References
+
+* https://en.wikipedia.org/wiki/Orientation_(vector_space)
+
+-/
+
+noncomputable theory
+
+section ordered_comm_ring
+
+variables {R : Type*} [ordered_comm_ring R]
+variables {M : Type*} [add_comm_group M] [module R M]
+variables {ι : Type*} [decidable_eq ι]
+
+namespace alternating_map
+
+/-- Two alternating maps have the same orientation if some positive multiples of them are
+equal (in the typical case over a field, this means each is a positive multiple of the other). -/
+def same_orientation (f g : alternating_map R M R ι) : Prop :=
+∃ (r₁ r₂ : R), 0 < r₁ ∧ 0 < r₂ ∧ r₁ • f = r₂ • g
```

Comment:
Is there a generalization of this for arbitrary modules that makes sense?
```suggestion
/-- Two elements of a module are <something> if some positive multiples of them are
equal (in the typical case over a field, this means each is a positive multiple of the other). -/
def smul_connected (f g : M) : Prop :=
∃ (r₁ r₂ : R), 0 < r₁ ∧ 0 < r₂ ∧ r₁ • f = r₂ • g
```

Reply:
The concept for an arbitrary module (over an `ordered_comm_semiring`, anyway; I don't know what makes sense in the non-commutative case, when this definition might not be an equivalence relation) is, geometrically, "on the same ray from the origin". (Which illustrates how this definition relates to one based on exterior powers: they're doing the same thing, but for rays in dual spaces.)

Reply:
Perhaps `same_orientation` is still a reasonable name for the general case then? Or maybe `same_direction`?

Reply:
Ah, think I found what I was looking for; I think this is a generalization of https://leanprover-community.github.io/mathlib_docs/group_theory/group_action/basic.html#mul_action.orbit_rel to monoids, applied to `pos_submonoid R`

Reply:
It's not obvious there is a unique useful generalization to monoids either. I used "some positive multiples are equal", because that easily gives an equivalence relation. Another variant would be "both are positive multiple of the same vector", but I'm not sure that's an equivalence relation in general. It would be possible to express either of those in terms of orbits for `pos_submonoid` (even if not in terms of `orbit_rel`), but it's not clear to me that you'd actually gain anything by expressing a definition that way.

Reply:
yeah, I agree with that. Perhaps let's just reference `orbit_rel` from the docstring, since the idea is similar. I think it would still be worth generalizing this new definition to arbitrary modules.

Reply:
> Another variant would be "both are positive multiple of the same vector"

I wonder if this would be better; I'm thinking of the example vectors `(1,2)` and `(2,4)` in the `ℤ`-module `zmod 4 × zmod 4`. Under the definition in the PR (generalized to an arbitrary module), these have the same orientation, but under your alternative they don't. I assume the problem appears for alternating maps too, and is just harder to construct?

Reply:
I'm now using `same_ray` for a general module over an `ordered_comm_semiring` (definition otherwise unchanged). The "both are positive multiple of the same vector" can fail transitivity for general modules over an `ordered_comm_semiring`; no idea whether it fails transitivity even when restricted to the module of alternating maps.

Reply:
And the doc string on `same_ray` mentions `mul_action.orbit_rel`.